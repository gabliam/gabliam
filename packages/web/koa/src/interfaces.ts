import { DefaultMethods } from '@gabliam/web-core';
import { koaRouter } from './koa';
/**
 * Represent a method that create an koa router
 */
export type RouterCreator = (path?: string) => koaRouter;

type Filter<T, U> = T extends U ? T : never;

export type KoaMethods = Filter<keyof koaRouter, DefaultMethods>;
