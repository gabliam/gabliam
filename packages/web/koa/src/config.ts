import { Bean, Joi, PluginConfig, Value } from '@gabliam/core';
import koaBody from 'koa-body';
import { KOA_PLUGIN_CONFIG } from './constants';

const koaBodyFormidableOptions = Joi.object()
  .keys({
    /**
     * {Integer} Limits the amount of memory all fields together (except files) can allocate in bytes.
     * If this value is exceeded, an 'error' event is emitted. The default size is 20MB.
     */
    maxFileSize: Joi.number().integer(),

    /**
     * {Integer} Limits the number of fields that the querystring parser will decode, default 1000
     */
    maxFields: Joi.number().integer(),

    /**
     * {Integer} Limits the amount of memory all fields together (except files) can allocate in bytes.
     * If this value is exceeded, an 'error' event is emitted, default 2mb (2 * 1024 * 1024)
     */
    maxFieldsSize: Joi.number().integer(),

    /**
     * {String} Sets the directory for placing file uploads in, default os.tmpDir()
     */
    uploadDir: Joi.string(),

    /**
     * {Boolean} Files written to uploadDir will include the extensions of the original files, default false
     */
    keepExtensions: Joi.bool(),

    /**
     * {String} If you want checksums calculated for incoming files, set this to either 'sha1' or 'md5', default false
     */
    hash: Joi.string(),

    /**
     * {Boolean} Multiple file uploads or no, default true
     */
    multiples: Joi.bool(),

    /**
     * {Function} Special callback on file begin. The function is executed directly by formidable.
     * It can be used to rename files before saving them to disk. See https://github.com/felixge/node-formidable#filebegin
     */
    onFileBegin: Joi.func(),
  })
  .default({ extended: false });

const koaBodyOptions = Joi.object().keys({
  /**
   * {Boolean} Patch request body to Node's ctx.req, default false
   *
   * Note: You can patch request body to Node or Koa in same time if you want.
   */
  patchNode: Joi.bool(),

  /**
   * {Boolean} Patch request body to Koa's ctx.request, default true
   *
   * Note: You can patch request body to Node or Koa in same time if you want.
   */
  patchKoa: Joi.bool(),

  /**
   * {String|Integer} The byte (if integer) limit of the JSON body, default 1mb
   */
  jsonLimit: [Joi.string(), Joi.number().integer()],

  /**
   * {String|Integer} The byte (if integer) limit of the form body, default 56kb
   */
  formLimit: [Joi.string(), Joi.number().integer()],

  /**
   * {String|Integer} The byte (if integer) limit of the text body, default 56kb
   */
  textLimit: [Joi.string(), Joi.number().integer()],

  /**
   * {String} Sets encoding for incoming form fields, default utf-8
   */
  encoding: Joi.string(),

  /**
   * {Boolean} Parse multipart bodies, default false
   */
  multipart: Joi.bool(),

  /**
   * {Boolean} Parse urlencoded bodies, default true
   */
  urlencoded: Joi.bool(),

  /**
   * {Boolean} Parse text bodies, default true
   */
  text: Joi.bool(),

  /**
   * {Boolean} Parse json bodies, default true
   */
  json: Joi.bool(),

  /**
   * Toggles co-body strict mode; if true, only parses arrays or objects, default true
   */
  jsonStrict: Joi.bool(),

  /**
   * {Object} Options to pass to the formidable multipart parser
   */
  formidable: koaBodyFormidableOptions,

  /**
   * {Function} Custom error handle, if throw an error, you can customize the response - onError(error, context), default will throw
   */
  onError: Joi.func(),

  /**
   * {Boolean} If enabled, don't parse GET, HEAD, DELETE requests, default true
   *
   * GET, HEAD, and DELETE requests have no defined semantics for the request body,
   * but this doesn't mean they may not be valid in certain use cases.
   * koa-body is strict by default
   *
   * see http://tools.ietf.org/html/draft-ietf-httpbis-p2-semantics-19#section-6.3
   */
  strict: Joi.bool(),
});

export interface KoaConfig {
  koaBodyOptions: koaBody.IKoaBodyOptions;
}

@PluginConfig()
export class KoaPluginConfig implements KoaConfig {
  @Value('application.web.koaBodyOptions', koaBodyOptions)
  koaBodyOptions: koaBody.IKoaBodyOptions;

  @Bean(KOA_PLUGIN_CONFIG)
  restConfig(): KoaConfig {
    return {
      koaBodyOptions: this.koaBodyOptions,
    };
  }
}
