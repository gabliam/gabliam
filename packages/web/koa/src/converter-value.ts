import {
  convertValueFn,
  ExecutionContext,
  GabContext,
  ResponseEntity,
} from '@gabliam/web-core';

export const converterValue: convertValueFn = (
  ctx: GabContext,
  execCtx: ExecutionContext,
  result: any
) => {
  const methodInfo = execCtx.getMethodInfo();
  const sendJsonValue = (value: any = '') => {
    let val: any;
    try {
      val = JSON.stringify(value);
    } catch {
      /* istanbul ignore next */
      val = value;
    }
    ctx.type = 'application/json';
    ctx.body = val;
  };
  // response handler if the result is a ResponseEntity
  function responseEntityHandler(value: ResponseEntity) {
    if (value.hasHeader()) {
      Object.keys(value.headers).forEach(k =>
        ctx.set(k, '' + value.headers[k])
      );
    }

    if (methodInfo.json) {
      sendJsonValue(value.body);
    } else {
      ctx.body = value.body;
    }
    ctx.status = value.status;
  }
  if (!ctx.headersSent) {
    if (result !== undefined) {
      if (result instanceof ResponseEntity) {
        responseEntityHandler(result);
      } else if (methodInfo.json) {
        sendJsonValue(result);
      } else {
        ctx.body = result;
      }

      if (methodInfo.json) {
        ctx.type = 'application/json';
      }
    }
  }
};
