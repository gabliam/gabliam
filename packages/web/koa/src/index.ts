import { KoaPlugin } from './koa-plugin';

export * from 'http-status-codes';
export * from './interfaces';
export * from './koa';
export * from './koa-converter';
export * from './koa-interceptor';
export * from './request';
export * from './response';
export { KoaPlugin as default };
