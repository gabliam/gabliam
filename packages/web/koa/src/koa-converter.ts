import {
  Container,
  InjectContainer,
  INJECT_CONTAINER_KEY,
  Service,
  toPromise,
} from '@gabliam/core';
import {
  createInterceptorResolver,
  getContext,
  getExtractArgs,
  isInterceptor,
  BadInterceptorError,
} from '@gabliam/web-core';
import { koa } from './koa';

/**
 * Class for convert interceptor to Koa Middleware
 */
@InjectContainer()
@Service()
export class KoaConverter {
  public interceptorToMiddleware(clazz: any): koa.Middleware {
    const container = <Container>(this as any)[INJECT_CONTAINER_KEY];
    const instance = createInterceptorResolver(container)(clazz);
    if (!isInterceptor(instance)) {
      throw new BadInterceptorError(instance);
    }

    const extractArgs = getExtractArgs(container, instance, 'intercept', true);

    return async (context: koa.Context, next: () => Promise<any>) => {
      let wasCalled = false;
      const nextWrap = async () => {
        wasCalled = true;
        await next();
      };

      const args = await extractArgs(getContext(context.req), null, nextWrap);
      await toPromise((instance['intercept'] as any)(...args));

      if (!wasCalled) {
        await next();
      }
    };
  }
}
