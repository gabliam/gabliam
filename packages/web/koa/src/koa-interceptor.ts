import { Service } from '@gabliam/core';
import {
  Context,
  GabContext,
  Interceptor,
  InterceptorConstructor,
  METADATA_KEY,
  Next,
  UseInterceptors,
  nextFn,
} from '@gabliam/web-core';
import { koaRouter } from './koa';

/**
 * Test if target is a valid interceptor
 * @param target any
 */
export const isValidInterceptor = (target: any) => {
  const meta = Reflect.getOwnMetadata(
    METADATA_KEY.specialInterceptor,
    target.constructor || target
  );

  return meta === undefined || meta === 'koa';
};

/**
 * KoaInterceptor decorator
 * class is an koa interceptor.
 * Must return koaRouter middleware
 */
const KoaInterceptor = () => (target: any) => {
  Reflect.defineMetadata(METADATA_KEY.specialInterceptor, 'koa', target);
};

/**
 * Convert a Koa router middleware to a gabliam interceptor
 */
export const toInterceptor = (
  mid: koaRouter.IMiddleware
): InterceptorConstructor => {
  const clazz: InterceptorConstructor = class implements Interceptor {
    async intercept(context: GabContext, next: nextFn) {
      const ctx: koaRouter.IRouterContext = context.state.context;
      return await mid(ctx, next);
    }
  };
  Context()(clazz.prototype, 'intercept', 0);
  Next()(clazz.prototype, 'intercept', 1);
  Service()(clazz);
  KoaInterceptor()(clazz);

  // metadata
  Reflect.defineMetadata('design:type', Function, clazz.prototype, 'intercept');
  Reflect.defineMetadata(
    'design:paramtypes',
    [Function, Function],
    clazz.prototype,
    'intercept'
  );
  Reflect.defineMetadata(
    'design:returntype',
    Promise,
    clazz.prototype,
    'intercept'
  );
  return clazz;
};

/**
 * Alias for evict to use: UseInterceptors(toInterceptor(koaMid))
 */
export const UseKoaInterceptors = (...mids: koaRouter.IMiddleware[]) =>
  UseInterceptors(...mids.map(m => toInterceptor(m)));
