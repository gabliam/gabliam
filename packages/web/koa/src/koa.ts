import koa from 'koa';
import koaRouter from 'koa-router';

export { koa, koaRouter };
