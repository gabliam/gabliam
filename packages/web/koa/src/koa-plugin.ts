import {
  Container,
  GabliamPlugin,
  Plugin,
  Registry,
  Scan,
  toPromise,
} from '@gabliam/core';
import {
  APP,
  cleanPath,
  compose,
  ExecutionContext,
  getContext,
  InterceptorInfo,
  REQUEST_LISTENER_CREATOR,
  RestMetadata,
  SERVER,
  WebConfiguration,
  WebPluginBase,
  WebConfigurationContructor,
} from '@gabliam/web-core';
import d from 'debug';
import http from 'http';
import { CUSTOM_ROUTER_CREATOR } from './constants';
import { converterValue } from './converter-value';
import { KoaMethods, RouterCreator } from './interfaces';
import { koa, koaRouter } from './koa';
import { isValidInterceptor } from './koa-interceptor';
import { addContextMiddleware, addMiddlewares } from './middleware';

const debug = d('Gabliam:Plugin:ExpressPlugin');

@Plugin('KoaPlugin')
@Scan()
export class KoaPlugin extends WebPluginBase<koa> implements GabliamPlugin {
  constructor(config?: Partial<WebConfigurationContructor<koa>>) {
    super(config);
  }

  bindApp(
    container: Container,
    registry: Registry,
    webConfiguration: WebConfiguration,
  ): void {
    container.bind(APP).toConstantValue(new koa());
    container.bind(REQUEST_LISTENER_CREATOR).toConstantValue(() => {
      const app = container.get<koa>(APP);
      app.silent = true;
      return app.callback();
    });

    webConfiguration.addwebConfig({
      instance: addMiddlewares,
      order: -2,
    });

    webConfiguration.addwebConfig({
      instance: addContextMiddleware,
      order: -1,
    });
  }

  async destroy(container: Container, registry: Registry) {
    await this.stop(container, registry);
  }

  async stop(container: Container, registry: Registry) {
    try {
      // server can be undefined (if start is not called)
      const server = container.get<http.Server>(SERVER);
      return new Promise<void>((resolve) => {
        server.close(() => resolve());
      });
    } catch (e) {}
  }

  async buildControllers(
    restMetadata: RestMetadata<KoaMethods>,
    container: Container,
  ) {
    // get the router creator
    let routerCreator: RouterCreator = (prefix?: string) =>
      new koaRouter({
        prefix,
      });
    try {
      routerCreator = container.get<RouterCreator>(CUSTOM_ROUTER_CREATOR);
    } catch (e) {}

    for (const [
      controllerId,
      { methods, controllerPath },
    ] of restMetadata.controllerInfo) {
      const controller = container.get<object>(controllerId);
      let routerPath: string | undefined = cleanPath(
        `${restMetadata.rootPath}${controllerPath}`,
      );

      if (routerPath === '/') {
        routerPath = undefined;
      }

      debug(`New route : "${routerPath}"`);
      const router = routerCreator(routerPath);

      for (const methodInfo of methods) {
        const execCtx = new ExecutionContext(controller, methodInfo);

        let methodMetadataPath = methodInfo.methodPath;
        if (methodMetadataPath[0] !== '/') {
          methodMetadataPath = '/' + methodMetadataPath;
        }

        const addJsonHandler = async (
          context: koaRouter.IRouterContext,
          next: () => Promise<any>,
        ) => {
          context.state.jsonHandler = methodInfo.json;
          await next();
        };

        const interceptors = methodInfo.interceptors.filter((i) =>
          isValidInterceptor(i.instance),
        );

        // create handler
        const handler = this.handlerFactory(execCtx, interceptors);

        // register handler in router
        router[methodInfo.method](methodMetadataPath, addJsonHandler, handler);
      }
      const app = container.get<koa>(APP);

      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore koa-router bad definition
      app.use(router.routes()).use(router.allowedMethods());
    }
  }

  private handlerFactory(
    execCtx: ExecutionContext,
    interceptors: InterceptorInfo[],
  ): koaRouter.IMiddleware {
    return async (
      context: koaRouter.IRouterContext,
      next: () => Promise<any>,
    ) => {
      const composeInterceptor = compose(interceptors, converterValue);
      const req = context.req;
      const ctx = getContext(req);
      const methodInfo = execCtx.getMethodInfo();
      const controller = execCtx.getClass();

      const callNext = async () => {
        const args = await methodInfo.extractArgs(ctx, execCtx, next);
        return await toPromise(controller[methodInfo.methodName](...args));
      };

      await composeInterceptor(ctx, execCtx, callNext);
    };
  }
}
