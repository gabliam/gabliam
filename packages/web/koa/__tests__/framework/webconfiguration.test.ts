import { Service } from '@gabliam/core';
import {
  Controller,
  Get,
  Pipe,
  RequestParam,
  Interceptor,
  Next,
} from '@gabliam/web-core';
import { WebPluginTest } from '@gabliam/web-core/src/testing';
import * as sinon from 'sinon';
import KoaPlugin, { toInterceptor } from '../../src';

let appTest: WebPluginTest;

afterEach(async () => {
  await appTest.destroy();
});

describe('globalPipes', () => {
  @Service()
  class A implements Pipe {
    transform(value: any) {
      return (value += 'a');
    }
  }

  @Service()
  class B implements Pipe {
    transform(value: any) {
      return (value += 'b');
    }
  }

  @Controller('/')
  class TestController {
    @Get('/:id')
    public getTest(@RequestParam('id') id: string) {
      return id;
    }
  }

  const spyA = sinon.spy(A.prototype, 'transform');
  const spyB = sinon.spy(B.prototype, 'transform');

  beforeEach(() => {
    spyA.resetHistory();
    spyB.resetHistory();
  });

  test('should call global-level pipe correctly', async () => {
    appTest = new WebPluginTest([
      new KoaPlugin({
        globalPipes: [A, B],
      }),
    ]);
    appTest.addClass(TestController);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .get('/lol')
      .expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
  });
});

describe('GlobalInterceptor', () => {
  let result: string;

  @Service()
  class A implements Interceptor {
    async intercept(@Next() next: () => Promise<any>) {
      result += 'a';
      await next();
    }
  }

  async function b(ctx: any, nextFunc: () => Promise<any>) {
    result += 'b';
    await nextFunc();
  }

  const B = toInterceptor(b);

  @Service()
  class C implements Interceptor {
    async intercept() {
      result += 'c';
    }
  }

  const spyA = sinon.spy(A.prototype, 'intercept');
  const spyB = sinon.spy(B.prototype, 'intercept');
  const spyC = sinon.spy(C.prototype, 'intercept');

  beforeEach(() => {
    result = '';
    spyA.resetHistory();
    spyB.resetHistory();
    spyC.resetHistory();
  });

  @Controller('/')
  class TestController {
    @Get('/')
    public getTest() {
      return 'GET';
    }
  }

  test('should call global-level interceptor correctly', async () => {
    appTest = new WebPluginTest([
      new KoaPlugin({
        globalInterceptors: [A, B, C],
      }),
    ]);

    appTest.addClass(TestController);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .get('/')
      .expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });
});
