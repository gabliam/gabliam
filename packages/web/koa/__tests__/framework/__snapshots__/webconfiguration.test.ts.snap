// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalInterceptor should call global-level interceptor correctly 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`GlobalInterceptor should call global-level interceptor correctly 2`] = `"abc"`;

exports[`globalPipes should call global-level pipe correctly 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "5",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "lolab",
}
`;
